core/src/muxing.rs

```diff
@@ -99,31 +104,13 @@ pub trait StreamMuxer {
     /// An error can be generated if the connection has been closed.
     fn poll_event(&self, cx: &mut Context<'_>) -> Poll<Result<StreamMuxerEvent, Self::Error>>;
 
-    /// Opens a new outgoing substream, and produces the equivalent to a future that will be
-    /// resolved when it becomes available.
-    ///
-    /// The API of `OutboundSubstream` is totally opaque, and the object can only be interfaced
-    /// through the methods on the `StreamMuxer` trait.
-    fn open_outbound(&self) -> Self::OutboundSubstream;
-
-    /// Polls the outbound substream.
-    ///
-    /// If `Pending` is returned, then the current task will be notified once the substream
-    /// is ready to be polled, similar to the API of `Future::poll()`.
-    /// However, for each individual outbound substream, only the latest task that was used to
-    /// call this method may be notified.
-    ///
-    /// May panic or produce an undefined result if an earlier polling of the same substream
-    /// returned `Ready` or `Err`.
-    fn poll_outbound(
```

Comment:
What if, as a user, I want to poll for outbound substreams, while at the same time signaling to the remote that I am not willing to accept any more inbound substreams? In other words, I do I exercise back-pressure on the number of inbound substreams?

I know this is not supported in mplex or yamux today. It is supported by QUIC and will be supported by https://github.com/libp2p/specs/pull/394.

Reply:
> while at the same time signaling to the remote that I am not willing to accept any more inbound substreams?

If this is a requirement, I'd suggest we make a function on `StreamMuxer` to trigger this behaviour.

In general though, I am a bit hesitant to do this. The original design of the `StreamMuxer` trait had numerous features that were completely unused (at least within the libp2p ecosystem that is known to me). Unless we use this functionality as part of `libp2p-swarm`, I'd refrain from adding it.

Reply:
> > while at the same time signaling to the remote that I am not willing to accept any more inbound substreams?
> 
> If this is a requirement, I'd suggest we make a function on `StreamMuxer` to trigger this behaviour.

This will be a requirement. It isn't today, due to rust-libp2p not yet implementing a sophisticated stream multiplexer. Will follow up with more details.

Reply:
Fair point! Can we agree to design this once we have/add a sophisticated stream multiplexer? I am not confident that I can add this functionality now in a way that will be useful once we need it!

Reply:
> I am not confident that I can add this functionality now in a way that will be useful once we need it!

Sounds good to me, especially since we don't prioritize outbound streams over inbound ones today.

https://github.com/libp2p/rust-libp2p/blob/3611d69e0911c5551270eaf8a6aa43d9b49f8782/swarm/src/connection/substream.rs#L154-L175